lib/matplotlib/tests/test_axes.py

```diff
@@ -7532,6 +7532,29 @@ def test_twinx_knows_limits():
     assert_array_equal(xtwin.viewLim.intervalx, ax2.viewLim.intervalx)
 
 
+class SubclassAxes(Axes):
+    def __init__(self, *args, foo, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.foo = foo
+
+
+@pytest.mark.parametrize("twinning", ["twinx", "twiny"])
+@pytest.mark.parametrize(("axes_class", "kw0", "kw1"), [
+    (Axes, {}, {}),
+    (SubclassAxes, {"foo": 0}, {"foo": 1}),
+])
+def test_twinning_subclass(twinning, axes_class, kw0, kw1):
+    fig = plt.figure()
+    classed_ax = fig.add_subplot(axes_class=axes_class, **kw0)
+    for k, v in kw0.items():
+        assert getattr(classed_ax, k) == v
+
+    twin = getattr(classed_ax, twinning)(axes_class=axes_class, **kw1)
+    assert type(twin) is axes_class
+    for k, v in kw1.items():
+        assert getattr(twin, k) == v
```

Comment:
I would write the tests like this.
- We want to particularly test that the types of the twins are not coupled. You need mixtures of Axes/SubclassAxes for that. That it works if both classes have the same `axes_class` specified is a much weaker assertion, and is almost certainly implied if the mixtures work.
- I'd remove the parametrization to make the code more readable. I think we can live without the kw-parametriztation. One can debate on wether the 3-line duplication (twinx/y() + 2 asserts) is worse or `twin = getattr(classed_ax, twinning)(axes_class=axes_class, **kw1)`.

```suggestion
def test_twinning_with_axes_class():
    """Check that twinx/y(axes_class=...) gives the appropriate class."""
    fig, ax = plt.subplots()
    twinx = ax.twinx(axes_class=SubclassAxes, foo=1)
    assert isinstance(twinx, SubclassAxes)
    assert twinx.foo == 1
    twiny = ax.twiny(axes_class=SubclassAxes, foo=2)
    assert isinstance(twinx, SubclassAxes)
    assert twinx.foo == 2


def test_twinning_default_axes_class():
    """Check that the default class for twinx/y() is Axes, even if the original is an Axes subclass."""
    fig, ax = plt.subplots(subplot_kw=dict(axes_class=SubclassAxes, foo=1))
    twinx = ax.twinx()
    assert type(twinx) is Axes
    twiny = ax.twiny()
    assert type(twiny) is Axes
```